Implement datetime_format() formula function

paljort opened the pull request:
# What is in this PR
This PR implements the `datetime_format()` formula function.

# How to test this PR
To test this PR, just enable the Advanced formula mode in any formula input field and type in the formula, e.g. `datetime_format(now(), 'DD/MM/YYYY HH:mm:ss', 'UTC')`

### Testing Frontend
To test the frontend changes, you can use Application Builder's preview mode which resolves the formula.

In AB, try creating a Heading element and provide the formula and you should it see resolved:
> <img width="777" height="228" alt="image" src="https://github.com/user-attachments/assets/360deb81-7106-4288-be7d-b83a8e2d6de9" />

### Testing Backend
To test the backend changes, you can use Automation Builder and test a node (e.g. Create Row):
> <img width="346" height="357" alt="image" src="https://github.com/user-attachments/assets/20c88f07-bbca-4aac-a1b1-3295696bebbd" />

Try testing different date/time formats.

# Checklist

- [ ] A changelog entry has been added to `changelog/entries/unreleased` using `changelog/src/changelog.py`
- [ ] New/updated **Premium/Enterprise features** are separated correctly in the premium or enterprise folder
- [ ] The latest **Chrome and Firefox** have been used to test any new frontend features
- [ ] [Documentation](https://github.com/baserow/baserow/blob/master/docs/index.md) has been updated
- [ ] [Quality Standards](https://github.com/baserow/baserow/blob/master/CONTRIBUTING.md#quality-standards) are met
- [ ] **Performance**: tables are still fast with 100k+ rows, 100+ field tables
- [ ] The [redoc API pages](https://api.baserow.io/api/redoc/) have been updated for any REST API changes
- [ ] Our [custom API docs](https://github.com/baserow/baserow/blob/master/web-frontend/modules/database/pages/APIDocsDatabase.vue) are updated for changes to endpoints accessed via API tokens
- [ ] The UI/UX has been updated following the [UI Style Guide](https://baserow.io/style-guide)
- [ ] Security impact of change has been considered

picklepete reviewed (changes requested):
Hey @paljort! I had fun testing this, it's going to make a lot of people happy in AB-land.

Two snags I ran into:

1. The timezone isn't actually optional, and that's just because we don't really have a way at the moment of making an arg optional. We can only override `validateNumberOfArgs` to see how many args we receive.
2. Moment's `tz` function cannot receive timezone abbreviations, so really users can only input UTC/GMT, which isn't ideal. I would imagine most non-technical users won't know their full timezone name...

I'm attaching a suggestion to address both of these:

[suggestion.patch](https://github.com/user-attachments/files/23305931/suggestion.patch)

We address (1) by allowing `BaserowRuntimeFormulaArgumentType` subclasses to set `optional:true`. This is then checked in `BaserowFormulaFunction.numArgs`, so we know what the number of required args are, and what the total number of args are (our min/max).

We address (2) by introducing `TimezoneBaserowRuntimeFormulaArgumentType`. It'll be optional in `RuntimeDateTimeFormat`. When it's tested we verify if we've got a timezone abbrevation, or the full name (so we don't allow users to store invalid TZ). When it's parsed, we convert an abbreviation into the full name (e.g. `PST` becomes `America/Los_Angeles`).

With these changes you can then try:

- `datetime_format(now(), 'DD-MM-YYYY HH:mm:ss', 'PST')`
- `datetime_format(now(), 'DD-MM-YYYY HH:mm:ss', 'America/Los_Angeles')`
- `datetime_format(now(), 'DD-MM-YYYY HH:mm:ss')` (your local time)

paljort commented:
> * Moment's `tz` function cannot receive timezone abbreviations, so really users can only input UTC/GMT, which isn't ideal. I would imagine most non-technical users won't know their full timezone name...

Hmm, I'm not sure if it's a good idea to allow abbreviations. I'm thinking about day light savings. E.g. `PST` isn't daylight savings aware. So if a user enters `PST`, what happens when daylight savings starts? Do we automatically convert it to `PDT`, or do we expect the user to remember to change it manually to `PDT`?

Another issue is that the abbreviation can be ambiguous. E.g. `IST` could refer to India, Ireland, or Israel. We could just support a small subset, as you suggested, but it could be perceived as inconsistent UX when another user wonders why their abbreviation doesn't work. So we're in a catch-22 here: a complete list would have the ambiguity problem, and an incomplete list would cause inconsistent UX.

I actually think it's acceptable to expect the user to provide the standard IANA timezone name. We could provide a link to a complete list of the timezone names in the help in the new formula input. What do you think?

paljort commented:
> The timezone isn't actually optional, and that's just because we don't really have a way at the moment of making an arg optional. We can only override `validateNumberOfArgs` to see how many args we receive.
> With these changes you can then try: `datetime_format(now(), 'DD-MM-YYYY HH:mm:ss') (your local time)`

I'm not sure it's a good idea to make the timezone optional. On the frontend we can assume/guess the user's local timezone, but we can't do that for the backend, e.g. an automation workflow gets executed in celery server-side, and we can only safely assume UTC.

But for certain use cases, like just displaying the current date/time, I can see it'd be useful to just "display local time".

Hmm, let me think on this more..

paljort commented:
> Might be worth poking @jrmi then, he wanted it to be optional too. I took the optional approach because in the backend code, there are statements for 2 and 3 args.

After speaking with @jrmi:
- I'll be implementing the optional timezone, in this MR. This will need some changes to send the timezone from the frontend -> backend.
- For the abbreviation, it's not a blocker. I'll log an issue for this and we can defer it for now.

paljort commented on backend/src/baserow/core/formula/argument_types.py:
```diff
@@ -87,3 +94,14 @@ def test(self, value):
+
+
+class TimezoneBaserowRuntimeFormulaArgumentType(BaserowRuntimeFormulaArgumentType):
+    def test(self, value):
```
Although we won't actually validate the argument type in the frontend (since we won't have the data for all types), this will still prevent invalid timezone names from being persisted.

paljort commented:
@picklepete Thanks for your review! I've made these changes:
- Added support for handling timezones via the user data provider in [6821b0c](https://github.com/baserow/baserow/pull/4148/commits/6821b0c5ff000876e2f906efd6e7ce59d85ffba8)
- Added a new Timezone formula arg type and made it optional in [c13fc61](https://github.com/baserow/baserow/pull/4148/commits/c13fc61bab87e2ed2517db33e74b32409b8910e4)

I also created two issues:
- [The Application settings should support setting a timezone](https://github.com/baserow/baserow/issues/4157)
- [Consider supporting Timezone Abbreviations for datetime formulas](https://github.com/baserow/baserow/issues/4158)

When you have a chance, could you please have a look?

paljort commented on backend/src/baserow/core/formula/argument_types.py:
```diff
@@ -87,3 +94,14 @@ def test(self, value):
+
+
+class TimezoneBaserowRuntimeFormulaArgumentType(BaserowRuntimeFormulaArgumentType):
+    def test(self, value):
```
Although we won't actually validate the argument type in the frontend (since we won't have the data for all types), this will still prevent invalid timezone names from being persisted.

paljort replied:
Sorry, to clarify, I meant that the arguments inside a function won't always be resolved, so we can't always validate them. E.g. in the case of `NumberBaserowRuntimeFormulaArgumentType`, we have a check for value being `undefined`, since for instance `get('previous_node.5.1')` will return undefined.
